hpy/tools/autogen/public_api.h

```diff
@@ -311,9 +311,23 @@ HPy_ID(168)
 int HPy_Is(HPyContext *ctx, HPy obj, HPy other);
 
 HPy_ID(169)
-void* HPy_AsStruct(HPyContext *ctx, HPy h);
+void* _HPy_AsStruct_Object(HPyContext *ctx, HPy h);
 HPy_ID(170)
-void* HPy_AsStructLegacy(HPyContext *ctx, HPy h);
+void* _HPy_AsStruct_Legacy(HPyContext *ctx, HPy h);
+HPy_ID(229)
+void* _HPy_AsStruct_Type(HPyContext *ctx, HPy h);
+HPy_ID(230)
+void* _HPy_AsStruct_Long(HPyContext *ctx, HPy h);
+HPy_ID(231)
+void* _HPy_AsStruct_Float(HPyContext *ctx, HPy h);
+HPy_ID(232)
+void* _HPy_AsStruct_Unicode(HPyContext *ctx, HPy h);
+HPy_ID(233)
+void* _HPy_AsStruct_Tuple(HPyContext *ctx, HPy h);
+HPy_ID(234)
+void* _HPy_AsStruct_List(HPyContext *ctx, HPy h);
+HPy_ID(235)
+void* _HPy_AsStruct_Slow(HPyContext *ctx, HPy h);
```

Comment:
Do we have a use-case for `_HPy_AsStruct_Slow`? If not then maybe we should not expose it yet?

Reply:
Hm, the debug context needs it. I see. The debug context does not have an access to the "internals" of the universal context it wraps, so we need to expose something for it. Ok, then we probably have no other option.

Reply:
yes, good point. So far we need something similar here, which is called by tp_traverse and tp_destroy.
https://github.com/hpyproject/hpy/blob/79f30c03cfaa4fdc3fa660b936a7ac172c3572aa/hpy/devel/src/runtime/ctx_type.c#L73-L81

+1 for keeping it private as long as there is no real need outside.

FWIW I just realized that in theory we could avoid it altogether by writing several versions of tp_traverse and tp_destroy, one for each shape. Not sure if it's worth it though (probably not)